src/backend/distributed/executor/multi_utility.c

```diff
@@ -883,6 +897,228 @@ ProcessAlterObjectSchemaStmt(AlterObjectSchemaStmt *alterObjectSchemaStmt,
 
 
 /*
+ * ProcessVacuumStmt processes vacuum statements that may need propagation to
+ * distributed tables. If a VACUUM or ANALYZE command references a distributed
+ * table, it is propagated to all involved nodes; otherwise, this function will
+ * immediately exit after some error checking.
+ *
+ * Unlike other Process functions within this file, this function does not
+ * return a modified parse node, as it is expected that the local VACUUM or
+ * ANALYZE has already been processed.
+ */
+static void
+ProcessVacuumStmt(VacuumStmt *vacuumStmt, const char *vacuumCommand)
+{
+	Oid relationId = InvalidOid;
+	List *taskList = NIL;
+	bool supportedVacuumStmt = false;
+
+	if (vacuumStmt->relation != NULL)
+	{
+		relationId = RangeVarGetRelid(vacuumStmt->relation, NoLock, false);
+	}
+
+	supportedVacuumStmt = IsSupportedDistributedVacuumStmt(relationId, vacuumStmt);
+	if (!supportedVacuumStmt)
+	{
+		return;
+	}
+
+	taskList = VacuumTaskList(relationId, vacuumStmt);
+
+	SavedMultiShardCommitProtocol = MultiShardCommitProtocol;
```

Comment:
I don't like this implementation of your approach. It means that we'll have a SavedMultiShardCommitProtocol set to some random old value if MultiShardCommitProtocol was changed since the last vacuum.